recommender/recommender.py

```diff
@@ -235,98 +246,119 @@ def error_handler(self, error_msg, event, resource_id=None):
-            self.recommendations[idx]['downvotes'] -= 1
-            result['newVotes'] = (self.recommendations[idx]['upvotes'] -
-                                  self.recommendations[idx]['downvotes'])
+        is_event_upvote = (data['event'] == 'recommender_upvote')
+        result['oldVotes'] = (self.recommendations[resource_id]['upvotes'] -
+                              self.recommendations[resource_id]['downvotes'])
+
+        if is_event_upvote:
+            if resource_id in self.upvoted_ids:
+                del self.upvoted_ids[self.upvoted_ids.index(resource_id)]
+                self.recommendations[resource_id]['upvotes'] -= 1
+                result['newVotes'] = (self.recommendations[resource_id]['upvotes'] -
+                                      self.recommendations[resource_id]['downvotes'])
+        else:
+            if resource_id in self.downvoted_ids:
+                del self.downvoted_ids[self.downvoted_ids.index(resource_id)]
+                self.recommendations[resource_id]['downvotes'] -= 1
+                result['newVotes'] = (self.recommendations[resource_id]['upvotes'] -
+                                      self.recommendations[resource_id]['downvotes'])
+
+        if 'newVotes' in result:
             result['Success'] = True
-            tracker.emit('recommender_downvote', result)
+            tracker.emit(data['event'], result)
             return result
 
-        if resource_id in self.upvoted_ids:
-            del self.upvoted_ids[self.upvoted_ids.index(resource_id)]
-            self.recommendations[idx]['upvotes'] -= 1
-            result['toggle'] = True
-        self.downvoted_ids.append(resource_id)
-        self.recommendations[idx]['downvotes'] += 1
-        result['newVotes'] = (self.recommendations[idx]['upvotes'] -
-                              self.recommendations[idx]['downvotes'])
+        if is_event_upvote:
+            if resource_id in self.downvoted_ids:
+                del self.downvoted_ids[self.downvoted_ids.index(resource_id)]
+                self.recommendations[resource_id]['downvotes'] -= 1
+                result['toggle'] = True
+            self.upvoted_ids.append(resource_id)
+            self.recommendations[resource_id]['upvotes'] += 1
+        else:
+            if resource_id in self.upvoted_ids:
+                del self.upvoted_ids[self.upvoted_ids.index(resource_id)]
+                self.recommendations[resource_id]['upvotes'] -= 1
+                result['toggle'] = True
+            self.downvoted_ids.append(resource_id)
+            self.recommendations[resource_id]['downvotes'] += 1
+
+        result['newVotes'] = (self.recommendations[resource_id]['upvotes'] -
+                              self.recommendations[resource_id]['downvotes'])
         result['Success'] = True
-        tracker.emit('recommender_downvote', result)
+        tracker.emit(data['event'], result)
         return result
```

Comment:
Is there ever a case where this function returns an error response? Seems like you're setting `result['Success'] = True` in many places. I don't think you need that (just use the HTTP status code to determine if the response was successful or not), but even if you do, you can set it in one place in the function, instead of several places.